examples/tls_client_session_wolfssl.cc

```diff
@@ -0,0 +1,163 @@
+ */
+#include "tls_client_session_wolfssl.h"
+
+#include <cassert>
+#include <iostream>
+
+#include "tls_client_context_wolfssl.h"
+#include "client_base.h"
+#include "template.h"
+#include "util.h"
+
+TLSClientSession::TLSClientSession() {}
+
+TLSClientSession::~TLSClientSession() {}
+
+extern Config config;
+
+namespace {
+int wolfssl_session_ticket_cb(WOLFSSL* ssl,
+                              const unsigned char* ticket, int ticketSz,
+                              void* cb_ctx) {
+  std::cerr << "session ticket calback invoked" << std::endl;
+  return 0;
+}
+} // namespace
+
+int TLSClientSession::init(bool &early_data_enabled,
+                           const TLSClientContext &tls_ctx,
+                           const char *remote_addr, ClientBase *client,
+                           uint32_t quic_version, AppProtocol app_proto) {
+  early_data_enabled = false;
+
+  auto ssl_ctx = tls_ctx.get_native_handle();
+
+  ssl_ = SSL_new(ssl_ctx);
+  if (!ssl_) {
+    std::cerr << "SSL_new: " << ERR_error_string(ERR_get_error(), nullptr)
+              << std::endl;
+    return -1;
+  }
+
+  SSL_set_app_data(ssl_, client->conn_ref());
+  SSL_set_connect_state(ssl_);
+
+  if ((quic_version & 0xff000000) == 0xff000000) {
+    wolfSSL_set_quic_use_legacy_codepoint(ssl_, 1);
+  } else {
+    wolfSSL_set_quic_use_legacy_codepoint(ssl_, 0);
+  }
+
+  switch (app_proto) {
+  case AppProtocol::H3: {
+    auto alpn = reinterpret_cast<const uint8_t *>(H3_ALPN);
+    auto alpnlen = str_size(H3_ALPN);
+    wolfSSL_set_alpn_protos(ssl_, alpn, alpnlen);
+    break;
+  }
+  case AppProtocol::HQ: {
+    auto alpn = reinterpret_cast<const uint8_t *>(HQ_ALPN);
+    auto alpnlen = str_size(HQ_ALPN);
+    wolfSSL_set_alpn_protos(ssl_, alpn, alpnlen);
+    break;
+  }
+  case AppProtocol::Perf:
+    /* TODO Not implemented yet */
+    assert(0);
+    break;
+  }
+
+  if (!config.sni.empty()) {
+    wolfSSL_UseSNI(ssl_, WOLFSSL_SNI_HOST_NAME,
+                   config.sni.data(), config.sni.length());
+  } else if (util::numeric_host(remote_addr)) {
+    // If remote host is numeric address, just send "localhost" as SNI
+    // for now.
+    wolfSSL_UseSNI(ssl_, WOLFSSL_SNI_HOST_NAME,
+                   "localhost", sizeof("localhost")-1);
+  } else {
+    wolfSSL_UseSNI(ssl_, WOLFSSL_SNI_HOST_NAME,
+                   remote_addr, strlen(remote_addr));
+  }
+
+  if (config.session_file) {
+#ifdef HAVE_SESSION_TICKET
+    auto f = BIO_new_file(config.session_file, "r");
+    if (f == nullptr) {
+      std::cerr << "Could not open TLS session file " << config.session_file
+                << std::endl;
+    } else {
+      unsigned char sbuffer[16*1024];
+      const unsigned char *pbuffer;
+      unsigned int sz = sizeof(sbuffer), ret;
+      WOLFSSL_SESSION *session;
+
+      sz = wolfSSL_BIO_read(f, sbuffer, sz);
+      if (sz <= 0) {
+        std::cerr << "Could not read TLS session file " << config.session_file
+                  << std::endl;
+      }
```

Comment:
It would be nice to add `else` so that the code path below is executed on failure.